File: Combination Sum II/Combination_Sum_II.py
# 52ms 95.29%
class Solution:
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res_list = []
        candidates.sort()
        
        def mysum(candidates, target, temp_list):
            for index in range(len(candidates)):
                if index > 0 and candidates[index] == candidates[index - 1]:
                    continue
                temp = target - candidates[index]
                if temp is 0:
                    res_list.append(temp_list + [candidates[index]])
                elif temp < 0:
                    break
                else:
                    mysum(candidates[index + 1:], temp, temp_list + [candidates[index]])
        
        mysum(candidates, target, [])
        return res_list

File: Combination Sum II/Combination_Sum_II.cpp
// 典型的回溯法
// Runtime: 8 ms, faster than 86.52% of C++ online submissions for Combination Sum II.
// Memory Usage: 8.8 MB, less than 94.33% of C++ online submissions for Combination Sum II.

class Solution 
{
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) 
    {
        // 边界条件处理
        sort(candidates.begin(), candidates.end());
        if (candidates.size() == 0 || candidates[0] > target)
            return vector<vector<int>>();
        
        vector<vector<int>> res;
        vector<int> temp;
        
        helper(res, candidates, temp, target, 0, 0);
        return res;
    }
    
private:
    void helper(vector<vector<int>>& res, vector<int>& candidates, vector<int>& temp, int target, int curSum, int start)
    {
        if (curSum == target)
        {
            res.push_back(temp);
            return ;
        }
        
        for (int i = start; i < candidates.size(); ++i)    
        {
            if (curSum + candidates[i] > target)
                break;
            if (i > start && candidates[i] == candidates[i - 1])
                continue;
            
            curSum += candidates[i];
            temp.push_back(candidates[i]);
            
            helper(res, candidates, temp, target, curSum, i + 1);
            
            curSum -= candidates[i];
            temp.pop_back();
        }
    }
};